### PR TITLE
Dynamic variables

### PR DESCRIPTION
**Pull request by Jade-GG**
With this update, we can write strings like `{billing.firstname} {billing.lastname}` and it will return the correct data for the user. Can also use custom_attributes data, like:

```
{custom:dynamics_company}
```

### CODE REVIEW

File: resources/js/callbacks.js [hunk 1]
```diff
@@ -19,20 +19,67 @@ Vue.prototype.transformAmCustomFormResponse = async function (response) {
...
+        let replaceVariable = (input) => {
+            switch(input) {
+                case '{product_url}':
+                case '{url}': return window.location.href
+
+                default:
+                    let parts = input.substring(1, input.length - 1).split('.')
+
+                    let pointer = window.app?.user
+                    if(!pointer) {
```
Comment (indykoning) [inline resources/js/callbacks.js]: Pointer seems like a very general name. It might be best to keep the name `user`

File: resources/js/callbacks.js [hunk 2]
```diff
@@ -19,20 +19,67 @@ Vue.prototype.transformAmCustomFormResponse = async function (response) {
...
+                            pointer = pointer.custom_attributes?.find(attribute => attribute.attribute_code == part.substring(7)).value
+                        } else {
+                            switch(part) {
+                                case 'address':
+                                    pointer = pointer.addresses[0]
+                                    break
+                                case 'billing':
+                                case 'shipping':
+                                    let id = pointer['default_'+part]
+                                    pointer = pointer.addresses.find(address => address.id == id)
```
Comment (indykoning) [inline resources/js/callbacks.js]:
```suggestion
                                    pointer = pointer.addresses.find(address => address.id === id)
```

File: resources/js/callbacks.js [hunk 3]
```diff
@@ -19,20 +19,67 @@ Vue.prototype.transformAmCustomFormResponse = async function (response) {
...
+
+                    let pointer = window.app?.user
+                    if(!pointer) {
+                        return ''
+                    }
+
+                    // Iterate through the parts to navigate the user object
+                    for(const part of parts) {
+                        if(part.startsWith('custom:')) {
+                            pointer = pointer.custom_attributes?.find(attribute => attribute.attribute_code == part.substring(7)).value
```
Comment (indykoning) [inline resources/js/callbacks.js]:
```suggestion
                            pointer = pointer.custom_attributes?.find(attribute => attribute.attribute_code === part.substring(7)).value
```

File: resources/js/callbacks.js [hunk 4]
```diff
@@ -19,20 +19,67 @@ Vue.prototype.transformAmCustomFormResponse = async function (response) {
...
+                        } else {
```
Comment (indykoning) [inline resources/js/callbacks.js]: If this is separated to another function we will not need this else, since we can return early in the if

File: resources/js/callbacks.js [hunk 5]
```diff
@@ -19,20 +19,67 @@ Vue.prototype.transformAmCustomFormResponse = async function (response) {
...
+        // Get all {}-enclosed variables and replace them
+        let matches = value.matchAll('{[^\}]*}')
+        for(const match of matches) {
+            let input = match[0]
+            if(!input) {
+                continue
+            }
+
+            value = value.replaceAll(input, replaceVariable(input))
+        }
```
Comment (indykoning) [inline resources/js/callbacks.js]: I suggest using groups so you don't need to use further string functions to unpack the variable names `(?<replace>{(?<variable>[^\}]*)})`
```
groups: 
    replace: "{testest}"
    variable: "testest"
```
```suggestion
        let matches = value.matchAll('(?<replace>{(?<variable>[^\}]*)})')
        for(const match of matches) {
            let input = match[0]
            if(!input) {
                continue
            }

            value = value.replaceAll(input.groups.replace, replaceVariable(input.groups.variable))
        }
```

**Review by indykoning — CHANGES_REQUESTED** [review]
Cool! 🚀  the amount of nesting makes it difficult to read and understand the code though. Could you spread them to more functions?

It also feels like i'm missing some of the variables described here https://amasty.com/docs/doku.php?id=magento_2:custom_form#:~:text=Magento%20Commerce%20Customer%20and%20Customer%20Address%20attributes%20as%20variables. Mainly the ones available within the address now only being available within the shipping or billing address now

File: resources/js/callbacks.js (hunk 5 above)
Comment (indykoning) [reply resources/js/callbacks.js]: Actually, looking at this, the for loop is not necessary
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replaceAll#:~:text=TypeError%20is%20thrown.-,replacement,-Can%20be%20a

it can be a string or a function
The function follows this
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replace#specifying_a_function_as_the_replacement

**Comment by Jade-GG** [comment]
> It also feels like i'm missing some of the variables described here

It's plausible, with this I only made sure that all of the variables previously supported by this package were also supported in this update.

File: resources/js/callbacks.js (hunk 1 above)
Comment (Jade-GG) [reply resources/js/callbacks.js]: `user` as a name wouldn't necessarily make sense though as it doesn't keep being the user object. It's used to navigate through objects, so `pointer` to me made sense. Something like `currentObject` could also work.

File: resources/js/callbacks.js (hunk 2 above)
Comment (Jade-GG) [reply resources/js/callbacks.js]: On this one we actually are comparing a string to an integer (`"2095" == 2095` is true but `"2095" === 2095` is false), so this one actually cannot be a strict equality comparison.

File: resources/js/callbacks.js [hunk 6]
```diff
@@ -19,20 +19,79 @@ Vue.prototype.transformAmCustomFormResponse = async function (response) {
...
+
+            // Return empty string when not defined
+            if(!pointer) {
+                if(fallback) {
+                    return fallback
+                }
+                return ''
+            }
+        }
+        return pointer
```
Comment (indykoning) [inline resources/js/callbacks.js]: I think it's best to move the fallback to the place where the pointer does not exist, so we're not needlessly traversing the object every time the fallback isn't needed
```suggestion
        for(const part of parts) {
            pointer = navigate(pointer, part)

            // Return empty string when not defined
            if(!pointer) {
                let fallback = navigate(pointer?.addresses?.[0], part)
                if(fallback) {
                    return fallback
                }
                return ''
            }
        }
        return pointer
```
or
```suggestion
        for(const part of parts) {
            pointer = navigate(pointer, part)

            if(pointer) {
                continue;
            }
            
            let fallback = navigate(pointer?.addresses?.[0], part)
            if(fallback) {
                return fallback
            }
            
            return ''
        }
        return pointer
```

**Comment by indykoning** [comment]
I've also found
```js
let obj = {'test': {'test2': {'test3': 'value'}}}

'test.test2.test3'.split('.').reduce((obj, index) => obj?.[index], obj)
// 'value'
```
Which might be useful for most cases, for the edge cases where we cannot perform something like this we probably do need to go the long way around.

Any incorrect key along the chain simply results in `undefined` being returned

File: resources/js/callbacks.js (hunk 6 above)
Comment (Jade-GG) [reply resources/js/callbacks.js]: Problem there is that by the time we get to a null pointer, we can't traverse backwards and get the addresses. I guess I could just go and reset the pointer to `this.$root.user?.addresses?.[0]` to save on needlessly traversing every time?

**Review by indykoning — APPROVED** [review]
